File: db_ets.erl
-module(db_ets).
-export([start_link/0, stop/0]).
-export([write/2, delete/1, read/1, match/1]).
-export([init/1, terminate/2, handle_call/3, handle_cast/2]).
-behavior(gen_server).


start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

stop() ->
    gen_server:cast(?MODULE, stop).

init(_Args) ->
    {ok, ets:new(db, [named_table, set])}.

terminate(_Reason, _LoopData) ->
    ets:delete(db).

write(Key, Element) ->
    gen_server:cast(?MODULE, {write, Key, Element}).

delete(Key) ->
    gen_server:cast(?MODULE, {delete, Key}).

read(Key) ->
    gen_server:call(?MODULE, {read, Key}).

match(Element) ->
    gen_server:call(?MODULE, {match, Element}).

handle_cast({write, Key, Element}, LoopData) ->
    ets:insert(db, {Key, Element}),
    {noreply, LoopData};

handle_cast({delete, Key}, LoopData) ->
    ets:delete(db, Key),
    {noreply, LoopData};

handle_cast(stop, LoopData) ->
    {stop, normal, LoopData}.

handle_call({read, Key}, _From, LoopData) ->
    Reply = case ets:lookup(db, Key) of
		[] -> {error, instance};
		[{Key, Value}] -> {ok, Value}
	    end,
    {reply, Reply, LoopData};

handle_call({match, Element}, _From, LoopData) ->
    Reply = lists:flatten(ets:match(db, {'$1', Element})),
    {reply, Reply, LoopData}.

File: db_ets/lib/db_ets/db_ets.ex
defmodule DbEts do
  @moduledoc """
  GenServer implementing simple database using ETS table.
  """
  use GenServer

  @table_name :db_ex

  def start_link(_opts) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
  end

  def stop do
    GenServer.cast(__MODULE__, :stop)
  end

  def write(key, element) do
    GenServer.cast(__MODULE__, {:write, key, element})
  end

  def delete(key) do
    GenServer.cast(__MODULE__, {:delete, key})
  end

  def read(key) do
    GenServer.call(__MODULE__, {:read, key})
  end

  def match(element) do
    GenServer.call(__MODULE__, {:match, element})
  end

  @impl true
  def init(_opts) do
    :ets.new(@table_name, [:named_table, :set])
    {:ok, []}
  end

  @impl true
  def terminate(_reason, _state) do
    :ets.delete(@table_name)
  end

  @impl true
  def handle_cast({:write, key, element}, state) do
    :ets.insert(@table_name, {key, element})
    :telemetry.execute([:db_ets, :records], %{count: 1}, %{})
    :telemetry.execute([:db_ets, :records, :write], %{count: 1}, %{})
    {:noreply, state}
  end

  @impl true
  def handle_cast({:delete, key}, state) do
    :ets.delete(@table_name, key)
    :telemetry.execute([:db_ets, :records], %{count: -1}, %{})
    {:noreply, state}
  end

  @impl true
  def handle_cast(:stop, state) do
    {:stop, :normal, state}
  end

  @impl true
  def handle_call({:read, key}, _from, state) do
    read_function =
      @table_name
      |> :ets.lookup(key)
      |> case do
        [] -> {:error, :not_found}
        [{^key, value}] -> {:ok, value}
      end

    reply = :telemetry.span([:db_ets, :records, :read], %{}, fn -> {read_function, %{}} end)

    {:reply, reply, state}
  end

  @impl true
  def handle_call({:match, element}, _from, state) do
    reply =
      @table_name
      |> :ets.match({:"$1", element})
      |> List.flatten()

    {:reply, reply, state}
  end
end
